File: php/tests/LiteSimpleConsumerTest.php
<?php
/**
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

namespace Apache\Rocketmq\Test;

require_once __DIR__ . '/TestRunner.php';
require_once __DIR__ . '/../LiteSimpleConsumer.php';
require_once __DIR__ . '/../Logger.php';

use Apache\Rocketmq\LiteSimpleConsumer;

/**
 * Tests for LiteSimpleConsumer validation rules.
 * Mirrors Java's LiteSimpleConsumerBuilderImplTest.
 */
class LiteSimpleConsumerTest
{
    public function setUp(): void
    {
        \Apache\Rocketmq\Logger::close();
    }

    /**
     * Mirrors Java: testSetConsumerGroupWithNull
     */
    public function testConstructorWithNullConsumerGroup()
    {
        TestRunner::assertThrows(\InvalidArgumentException::class, function() {
            new LiteSimpleConsumer('127.0.0.1:9876', '', 'parent-topic');
        }, "Empty consumer group should throw");
    }

    /**
     * Mirrors Java: testBindTopicWithNull
     */
    public function testConstructorWithEmptyParentTopic()
    {
        TestRunner::assertThrows(\InvalidArgumentException::class, function() {
            new LiteSimpleConsumer('127.0.0.1:9876', 'test-group', '');
        }, "Empty parent topic should throw");
    }

    /**
     * Mirrors Java: testBindTopicWithBlank
     */
    public function testConstructorWithWhitespaceParentTopic()
    {
        TestRunner::assertThrows(\InvalidArgumentException::class, function() {
            new LiteSimpleConsumer('127.0.0.1:9876', 'test-group', '  ');
        }, "Whitespace-only parent topic should throw");
    }

    /**
     * Mirrors Java: testBindTopicWithBlankString
     */
    public function testSubscribeLiteWithEmptyTopic()
    {
        $consumer = new LiteSimpleConsumer('127.0.0.1:9876', 'test-group', 'parent-topic');

        TestRunner::assertThrows(\InvalidArgumentException::class, function() use ($consumer) {
            $consumer->subscribeLite('');
        }, "Empty lite topic should throw");
    }

    /**
     * Mirrors Java: testBindTopicWithWhitespaces
     */
    public function testSubscribeLiteWithWhitespaceTopic()
    {
        $consumer = new LiteSimpleConsumer('127.0.0.1:9876', 'test-group', 'parent-topic');

        TestRunner::assertThrows(\InvalidArgumentException::class, function() use ($consumer) {
            $consumer->subscribeLite('   ');
        }, "Whitespace-only lite topic should throw");
    }

    /**
     * Mirrors Java: testSetNegativeMaxCacheMessageCount
     */
    public function testConstructorWithZeroAwaitDuration()
    {
        $consumer = new LiteSimpleConsumer('127.0.0.1:9876', 'test-group', 'parent-topic', [
            'awaitDuration' => 0,
        ]);

        $ref = new \ReflectionProperty($consumer, 'awaitDuration');
        $ref->setAccessible(true);
        $actual = $ref->getValue($consumer);
        TestRunner::assertEquals(0, $actual, "awaitDuration should be 0");
    }

    /**
     * Mirrors Java: testBuildWithoutConsumerGroup
     */
    public function testStartWithoutLiteTopics()
    {
        $consumer = new LiteSimpleConsumer('127.0.0.1:9876', 'test-group', 'parent-topic');

        TestRunner::assertThrows(\RuntimeException::class, function() use ($consumer) {
            $consumer->start();
        }, "Start without lite topics should throw");
    }

    /**
     * Tests that subscribeLite works before start.
     */
    public function testSubscribeLiteBeforeStart()
    {
        $consumer = new LiteSimpleConsumer('127.0.0.1:9876', 'test-group', 'parent-topic');

        $consumer->subscribeLite('lite-topic-1');
        $consumer->subscribeLite('lite-topic-2', 'tagA');

        $topics = $consumer->getLiteTopics();
        TestRunner::assertEquals(
            2,
            count($topics),
            "Should have 2 lite topics"
        );
        TestRunner::assertTrue(
            in_array('lite-topic-1', $topics),
            "lite-topic-1 should be in topics"
        );
        TestRunner::assertTrue(
            in_array('lite-topic-2', $topics),
            "lite-topic-2 should be in topics"
        );
    }

    /**
     * Tests unsubscribeLite before start.
     */
    public function testUnsubscribeLiteBeforeStart()
    {
        $consumer = new LiteSimpleConsumer('127.0.0.1:9876', 'test-group', 'parent-topic');

        $consumer->subscribeLite('lite-topic');
        $consumer->unsubscribeLite('lite-topic');

        $topics = $consumer->getLiteTopics();
        TestRunner::assertTrue(
            empty($topics),
            "Lite topics should be empty after unsubscribe"
        );
    }

    /**
     * Tests receive() before start throws.
     */
    public function testReceiveWithoutStart()
    {
        $consumer = new LiteSimpleConsumer('127.0.0.1:9876', 'test-group', 'parent-topic');

        TestRunner::assertThrows(\RuntimeException::class, function() use ($consumer) {
            $consumer->receive(10);
        }, "Receive before start should throw");
    }

    /**
     * Tests ack() before start throws.
     */
    public function testAckWithoutStart()
    {
        $consumer = new LiteSimpleConsumer('127.0.0.1:9876', 'test-group', 'parent-topic');

        TestRunner::assertThrows(\RuntimeException::class, function() use ($consumer) {
            $consumer->ack(new \stdClass());
        }, "Ack before start should throw");
    }

    /**
     * Tests that start() returns silently when already running.
     */
    public function testStartWhenAlreadyRunning()
    {
        $consumer = new LiteSimpleConsumer('127.0.0.1:9876', 'test-group', 'parent-topic');
        $consumer->subscribeLite('lite-topic');

        $ref = new \ReflectionProperty($consumer, 'isRunning');
        $ref->setAccessible(true);
        $ref->setValue($consumer, true);

        $consumer->start();
        TestRunner::assertTrue($consumer->isRunning(), "Consumer should still be running");
    }

    /**
     * Tests that lite topic name length is validated.
     */
    public function testLiteTopicNameTooLong()
    {
        $consumer = new LiteSimpleConsumer('127.0.0.1:9876', 'test-group', 'parent-topic', [
            'maxLiteTopicSize' => 10,
        ]);

        TestRunner::assertThrows(\RuntimeException::class, function() use ($consumer) {
            $consumer->subscribeLite('this-is-a-very-long-lite-topic-name');
        }, "Long lite topic name should throw");
    }

    /**
     * Mirrors Java: testMethodChaining
     */
    public function testMethodChaining()
    {
        $consumer = new LiteSimpleConsumer('127.0.0.1:9876', 'test-group', 'parent-topic');

        $result = $consumer->subscribeLite('topic1');
        TestRunner::assertTrue(
            $result === $consumer,
            "subscribeLite should return \$this for chaining"
        );

        $result = $consumer->unsubscribeLite('topic1');
        TestRunner::assertTrue(
            $result === $consumer,
            "unsubscribeLite should return \$this for chaining"
        );
    }

    /**
     * Tests getClientId returns expected format.
     */
    public function testGetClientId()
    {
        $consumer = new LiteSimpleConsumer('127.0.0.1:9876', 'test-group', 'parent-topic');
        $clientId = $consumer->getClientId();

        TestRunner::assertTrue(
            strlen($clientId) > 0,
            "ClientId should be non-empty"
        );
        TestRunner::assertTrue(
            strpos($clientId, 'php-lite-simple-consumer-') === 0,
            "ClientId should start with expected prefix"
        );
    }
}

echo "=== LiteSimpleConsumerTest ===\n";
TestRunner::run(new LiteSimpleConsumerTest());

File: rust/tests/lite_simple_consumer_test.rs
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

//! Offline tests for LiteSimpleConsumer.
//!
//! These tests run without any RocketMQ deployment: they only assert the public API contract
//! (validation, accessors, life cycle checks) of `LiteSimpleConsumer`.

#[cfg(test)]
mod lite_simple_consumer_tests {
    use rocketmq::conf::{ClientOption, SimpleConsumerOption};
    use rocketmq::error::ErrorKind;
    use rocketmq::model::offset_option::{OffsetOption, OffsetPolicy};
    use rocketmq::{LiteSimpleConsumer, LiteSimpleConsumerTrait};

    fn client_option() -> ClientOption {
        let mut client_option = ClientOption::default();
        client_option.set_access_url("http://localhost:8080");
        client_option
    }

    fn consumer_option(consumer_group: &str) -> SimpleConsumerOption {
        let mut option = SimpleConsumerOption::default();
        option.set_consumer_group(consumer_group);
        option
    }

    #[test]
    fn test_new_requires_consumer_group() {
        let result = LiteSimpleConsumer::new(
            client_option(),
            consumer_option(""),
            "parent_topic".to_string(),
        );
        let err = result.err().expect("consumer group is mandatory");
        assert_eq!(*err.kind(), ErrorKind::Config);
        assert_eq!(err.operation(), "lite_simple_consumer.new");
    }

    #[test]
    fn test_new_requires_bind_topic() {
        let result =
            LiteSimpleConsumer::new(client_option(), consumer_option("group"), "".to_string());
        let err = result.err().expect("bind topic is mandatory");
        assert_eq!(*err.kind(), ErrorKind::Config);
        assert!(err.message().contains("bind topic"));
    }

    #[test]
    fn test_new_succeeds_and_exposes_metadata() {
        let consumer = LiteSimpleConsumer::new(
            client_option(),
            consumer_option("test_group"),
            "parent_topic".to_string(),
        )
        .expect("consumer should be created");
        assert_eq!(consumer.get_consumer_group(), "test_group");
        assert_eq!(consumer.get_bind_topic(), "parent_topic");
        assert!(consumer.get_lite_topic_set().is_empty());
    }

    #[tokio::test]
    async fn test_subscribe_lite_before_start_fails() {
        let consumer = LiteSimpleConsumer::new(
            client_option(),
            consumer_option("test_group"),
            "parent_topic".to_string(),
        )
        .expect("consumer should be created");

        let err = consumer
            .subscribe_lite("lite_topic".to_string())
            .await
            .err()
            .expect("subscribeLite requires a started consumer");
        assert_eq!(*err.kind(), ErrorKind::ClientIsNotRunning);

        let err = consumer
            .subscribe_lite_with_offset(
                "lite_topic".to_string(),
                OffsetOption::from_policy(OffsetPolicy::Min),
            )
            .await
            .err()
            .expect("subscribeLite requires a started consumer");
        assert_eq!(*err.kind(), ErrorKind::ClientIsNotRunning);

        let err = consumer
            .unsubscribe_lite("lite_topic".to_string())
            .await
            .err()
            .expect("unsubscribeLite requires a started consumer");
        assert_eq!(*err.kind(), ErrorKind::ClientIsNotRunning);
    }
}
